### PR TITLE
fix snapshot read model

### LINKED ISSUE #30: Events are applied multiple times when fetching aggregate through snapshot store
I have recreated an environment to reproduce this issue.
You will find it here: https://github.com/gquemener/prooph-snapshotter-issue-poc.

Basically, the snapshotter applies the events multiple times on aggregate root.

I haven't yet detected the root cause of this issue, but I'm suspecting something around the `persistBlockSize` of the `PdoEventStoreProjector` as the issue seems to appear only when the event stream contain more than 1000 events (1000 being the default `persistBlockSize`).

### PR DESCRIPTION
**Pull request by prolic**
resolves https://github.com/prooph/snapshotter/issues/30

/cc @gquemener

### CODE REVIEW

**Comment by prolic** [comment]
I will add official php 7.2 support, then we can release new version.

File: src/StreamSnapshotProjection.php [hunk 1]
```diff
@@ -31,15 +31,16 @@ public function __construct(ReadModelProjector $readModelProjector, string $stre
...
         $this->streamName = $streamName;
-    }
 
-    public function __invoke(bool $keepRunning = true)
-    {
         $this->readModelProjector
             ->fromStream($this->streamName)
             ->whenAny(function ($state, Message $event): void {
                 $this->readModel()->stack('replay', $event);
-            })
```
Comment (gquemener) [inline src/StreamSnapshotProjection.php]: Any special reason why moving this logic into the ctor?

There's no need to store the `streamName` if you do so.

You could then do:

```php
  1 <?php                                                                                                                   
  2                                                                                                                         
  3 public function __construct(ReadModelProjector $readModelProjector, string $streamName)                                                                                                                                                   
  4 {                                                                                                                       
  5     $this->readModelProjector = $readModelProjector                                                                     
  6         ->fromStream($streamName)                                                                                       
  7         ->whenAny(function ($state, Message $event): void {                                                             
  8             $this->readModel()->stack('replay', $event);                                                                
  9         });                                                                                                             
 10 }         
```

File: src/StreamSnapshotProjection.php [hunk 2]
```diff
@@ -31,15 +31,16 @@ public function __construct(ReadModelProjector $readModelProjector, string $stre
...
-            ->run($keepRunning);
+            });
+    }
+
+    public function __invoke(bool $keepRunning = true)
```
Comment (gquemener) [inline src/StreamSnapshotProjection.php]: I'm also curious about why using `__invoke` here?

By the way, the doc is wrong then (it mentions a `run()` method which does not exist: https://github.com/prooph/snapshotter/blame/master/docs/snapshots.md#L53)

**Comment by prolic** [comment]
Calling run() multiple times doesn't work. But good catch. I'm going to
clean up the constructor.

**Comment by gquemener** [comment]
Have you had time to consider [this comment](https://github.com/prooph/snapshotter/issues/30#issuecomment-351844856) @prolic ?

EDIT: The version check I suggest could instead be applied in the users' `apply` method (which will become the user's reponsability and will require no change in `prooph/event-sourcing`). I am not sure it's a good practice though.

**Comment by prolic** [comment]
@gquemener this can be added as additional check, just to make sure. So you can create a PR if you want to.

File: src/StreamSnapshotProjection.php (hunk 2 above)
Comment (prolic) [reply src/StreamSnapshotProjection.php]: updated the docs

**Comment by prolic** [comment]
@gquemener can you confirm as well the bug is fixed?

**Review by codeliner — APPROVED** [review]

**Comment by codeliner** [comment]
```
Repository without snapshot store
1006 events were applied 1 times.

Repository with snapshot store
1006 events were applied 1 times.
```

@prolic looks good

**Comment by gquemener** [comment]
Sorry for the delay, I confirm that it fixes the issue @prolic.

Thanks @codeliner.

**Pull request merged**